### PR TITLE
Fix RSAPSS salt length conversion

### PR DESCRIPTION
**Pull request by qmuntal**
This is the CNG counterpart of https://github.com/microsoft/go-crypto-openssl/pull/37. See that PR for detailed context.

### CODE REVIEW

File: cng/rsa.go [hunk 1]
```diff
@@ -325,17 +325,36 @@ func keyVerify(pkey bcrypt.KEY_HANDLE, info unsafe.Pointer, hashed, sig []byte,
...
 	if hashID == "" {
 		return info, errors.New("crypto/rsa: unsupported hash function")
 	}
 	info.AlgId = utf16PtrFromString(hashID)
+
+	// A salt length of -1 and 0 are valid Go sentinel values.
+	if saltLen <= -2 {
+		return info, errors.New("crypto/rsa: PSSOptions.SaltLength cannot be negative")
+	}
+	// CNG does not support salt length special cases like Go do,
```
Comment (dagood) [inline cng/rsa.go:339]:
```suggestion
	// CNG does not support salt length special cases like Go crypto does,
```

File: cng/rsa.go [hunk 2]
```diff
@@ -325,17 +325,36 @@ func keyVerify(pkey bcrypt.KEY_HANDLE, info unsafe.Pointer, hashed, sig []byte,
...
 	switch saltLen {
 	case -1: // rsa.PSSSaltLengthEqualsHash
 		info.Salt = uint32(h.Size())
 	case 0: // rsa.PSSSaltLengthAuto
-		err = errors.New("crypto/rsa: rsa.PSSSaltLengthAuto not supported")
+		if sign {
+			// Go sets the salt as large as possible when signing.
+			bits, err := getUint32(bcrypt.HANDLE(pkey), bcrypt.KEY_LENGTH)
+			if err != nil {
+				return info, errors.New("crypto/rsa: key length can't be retrieved" + err.Error())
```
Comment (dagood) [inline cng/rsa.go:349]: Space to keep the inner error readable?

```suggestion
				return info, errors.New("crypto/rsa: key length can't be retrieved " + err.Error())
```

File: cng/rsa.go [hunk 3]
```diff
@@ -325,17 +325,36 @@ func keyVerify(pkey bcrypt.KEY_HANDLE, info unsafe.Pointer, hashed, sig []byte,
...
+			}
+			info.Salt = (bits-1+7)/8 - 2 - uint32(h.Size())
```
Comment (dagood) [inline cng/rsa.go:351]: Why is this the largest salt? Seems something like "as much space as we have left in the key", but the constants don't seem obvious...

Comment (qmuntal) [reply cng/rsa.go]: I've added some comments to clarify this algorithm

**Review by dagood — APPROVED** [review]
Makes sense, thanks!

**Review by chsienki — APPROVED** [review]

**Pull request merged**